src/System.Runtime/tests/System/LazyTests.cs

```diff
@@ -103,17 +107,284 @@ public static void ToString_DoesntForceAllocation()
+
+            // When used with LazyThreadSafetyMode.PublicationOnly this causes a stack overflow
+            // Value_Invalid_Impl(ref x, new Lazy<int>(f, LazyThreadSafetyMode.PublicationOnly));
+        }
+        
+        public class InitiallyExceptionThrowingCtor
+        {
+            public static int counter = 0;
+            public static int getValue()
+            {
+                if (++counter < 5)
+                    throw new Exception();
+                else
+                    return counter;
+            }
+
+            public int Value { get; }
+
+            public InitiallyExceptionThrowingCtor()
+            {
+                Value = getValue();
+            }
+        }
+
+        public static IEnumerable<object[]> Ctor_ExceptionRecovery_MemberData()
+        {
+            yield return new object[] { new Lazy<InitiallyExceptionThrowingCtor>(), 5 };
+            yield return new object[] { new Lazy<InitiallyExceptionThrowingCtor>(true), 5 };
+            yield return new object[] { new Lazy<InitiallyExceptionThrowingCtor>(false), 5 };
+            yield return new object[] { new Lazy<InitiallyExceptionThrowingCtor>(LazyThreadSafetyMode.ExecutionAndPublication), 5 };
+            yield return new object[] { new Lazy<InitiallyExceptionThrowingCtor>(LazyThreadSafetyMode.None), 5 };
+            yield return new object[] { new Lazy<InitiallyExceptionThrowingCtor>(LazyThreadSafetyMode.PublicationOnly), 5 };
+        }
+
+        [Theory]
+        [MemberData(nameof(Ctor_ExceptionRecovery_MemberData))]
+        public static void Ctor_ExceptionRecovery(Lazy<InitiallyExceptionThrowingCtor> lazy, int? expected)
+        {
+            InitiallyExceptionThrowingCtor.counter = 0;
+            var result = default(InitiallyExceptionThrowingCtor);
+            for (var i = 0; i < 10; ++i)
+            {
+                try { result = lazy.Value; } catch (Exception) { }
+            }
+            if (expected == null)
+                Assert.Null(result);
+            else
+                Assert.Equal(result.Value, expected.Value);
+        }
+
+        private static void Value_ExceptionRecovery_IntImpl(Lazy<int> lazy, ref int counter, int expected)
+        {
+            counter = 0;
+            var result = 0;
+            for (var i = 0; i < 10; ++i)
+            {
+                try { result = lazy.Value; } catch (Exception) { }
+            }
+            Assert.Equal(result, expected);
+        }
+
+        private static void Value_ExceptionRecovery_StringImpl(Lazy<string> lazy, ref int counter, string expected)
+        {
+            counter = 0;
+            var result = default(string);
+            for (var i = 0; i < 10; ++i)
+            {
+                try { result = lazy.Value; } catch (Exception) { }
+            }
+            if (expected == null)
+                Assert.Null(result);
+            else
+                Assert.Equal(result, expected);
```

Comment:
Also note that expected should be the first argument to Equal, here and elsewhere